File: src/api/requirement.ts
import { invoke } from "@tauri-apps/api/core";
import {
  createWorkItem,
  deleteWorkItem,
  getWorkItemDetail,
  listParentProjects,
  listWorkItems,
  updateWorkItem,
} from "@/api/work-item";
import type { WorkItem } from "@/types/work-item";
import type {
  CreateRequirementPayload,
  RequirementDetail,
  RequirementItem,
  RequirementListQuery,
  RequirementListResult,
  UpdateRequirementDetailPayload,
  UpdateRequirementPayload,
} from "@/types/requirement";

export async function pingRequirement(): Promise<string> {
  return invoke<string>("requirement_ping");
}

export async function listRequirements(
  query: RequirementListQuery,
): Promise<RequirementListResult> {
  const result = await listWorkItems({
    page: query.page,
    pageSize: query.pageSize,
    kind: "requirement",
    parentId: query.projectId,
    keyword: query.keyword,
    status: query.status,
    priority: undefined,
    sortField: query.sortField === "updatedAt" ? "updatedAt" : undefined,
    sortOrder: query.sortOrder,
  });
  return {
    ...result,
    items: result.items.map(toRequirementItem),
  };
}

export async function listRequirementOwners(): Promise<string[]> {
  const result = await listWorkItems({
    page: 1,
    pageSize: 500,
    kind: "requirement",
  });
  return [...new Set(result.items.map((item) => item.owner).filter(Boolean))];
}

export async function createRequirement(
  payload: CreateRequirementPayload,
): Promise<RequirementItem> {
  const item = await createWorkItem({
    kind: "requirement",
    parentId: payload.projectId,
    title: payload.title,
    status: payload.status,
    priority: payload.priority,
    owner: payload.owner,
    effort: payload.effort,
    planMonth: payload.planMonth,
  });
  return toRequirementItem(item);
}

export async function updateRequirement(
  payload: UpdateRequirementPayload,
): Promise<RequirementItem> {
  const item = await updateWorkItem({
    id: payload.id,
    parentId: payload.projectId,
    title: payload.title,
    status: payload.status,
    priority: payload.priority,
    owner: payload.owner,
    effort: payload.effort,
    planMonth: payload.planMonth,
    content: payload.content,
  });
  return toRequirementItem(item);
}

export async function deleteRequirement(id: number): Promise<void> {
  return deleteWorkItem(id);
}

export async function getRequirementDetail(id: number): Promise<RequirementDetail> {
  const item = await getWorkItemDetail(id);
  return toRequirementItem(item) as RequirementDetail;
}

export async function updateRequirementDetail(
  payload: UpdateRequirementDetailPayload,
): Promise<RequirementItem> {
  return updateRequirement(payload);
}

export async function listRequirementProjects(): Promise<Array<{ id: number; projId: string; title: string }>> {
  const rows = await listParentProjects();
  return rows.map((row) => ({ id: row.id, projId: row.itemId, title: row.title }));
}

function toRequirementItem(item: WorkItem): RequirementItem {
  return {
    id: item.id,
    reqId: item.itemId,
    projectId: item.parentId,
    title: item.title,
    status: item.status,
    priority: item.priority,
    owner: item.owner,
    effort: item.effort ?? 0,
    planMonth: item.planMonth ?? "",
    content: item.content,
    updatedAt: item.updatedAt,
  };
}

File: src-tauri/src/commands/requirement.rs
use crate::services;
use crate::state::AppState;

#[tauri::command]
pub fn requirement_ping() -> &'static str {
    "requirement"
}

#[tauri::command]
pub async fn list_requirements(
    state: tauri::State<'_, AppState>,
    query: services::requirement::RequirementListQuery,
) -> Result<services::requirement::RequirementListResult, String> {
    services::requirement::list_requirements(&state.db, query)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_requirement_owners(state: tauri::State<'_, AppState>) -> Result<Vec<String>, String> {
    services::requirement::list_requirement_owners(&state.db)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_requirement(
    state: tauri::State<'_, AppState>,
    payload: services::requirement::CreateRequirementPayload,
) -> Result<services::requirement::RequirementDto, String> {
    services::requirement::create_requirement(&state.db, payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_requirement(
    state: tauri::State<'_, AppState>,
    payload: services::requirement::UpdateRequirementPayload,
) -> Result<services::requirement::RequirementDto, String> {
    services::requirement::update_requirement(&state.db, payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_requirement(state: tauri::State<'_, AppState>, id: i32) -> Result<(), String> {
    services::requirement::delete_requirement(&state.db, id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_requirement_detail(
    state: tauri::State<'_, AppState>,
    id: i32,
) -> Result<services::requirement::RequirementDto, String> {
    services::requirement::get_requirement_detail(&state.db, id)
        .await
        .map_err(|e| e.to_string())
}
